app/src/main/java/net/gsantner/markor/frontend/textview/HighlightingEditor.java

```diff
@@ -95,6 +107,66 @@ public void afterTextChanged(final Editable s) {
         setEmojiCompatEnabled(false);
     }
 
+    @Override
+    protected void onFinishInflate() {
+        super.onFinishInflate();
+        _defaultPaddingLeft = getPaddingLeft();
+    }
+
+    @Override
+    protected void onDraw(Canvas canvas) {
+        super.onDraw(canvas);
+
+        // If line numbers enabled
+        if (_nuEnabled) {
+            drawLineNumbers(canvas);
+        } else if (getPaddingLeft() != _defaultPaddingLeft) {
+            // Reset padding without line numbers fence
+            setPadding(_defaultPaddingLeft, getPaddingTop(), getPaddingRight(), getPaddingBottom());
+        }
+    }
+
+    private void drawLineNumbers(Canvas canvas) {
+        final int firstBaselineToTopHeight = getPaddingTop() - getPaint().getFontMetricsInt().top;
+        int number = 1;
+
+        _paint.setColor(Color.GRAY);
+        // Draw first line number
+        canvas.drawText(String.valueOf(number), LINE_NUMBERS_PADDING_LEFT, firstBaselineToTopHeight, _paint);
+        // Draw others line number
+        float y;
+        final float offsetY = firstBaselineToTopHeight - getTextSize() * 0.14f;
+        int top = Integer.MIN_VALUE;
+        int bottom = Integer.MAX_VALUE;
+        if (getParent() instanceof ScrollView) {
+            ScrollView scrollView = (ScrollView) getParent();
+            // Set range of current visible area, drawing line number only near the visible area
+            top = scrollView.getScrollY() - 100;
+            bottom = scrollView.getScrollY() + scrollView.getHeight();
+        }
+
+        final int count = getLineCount();
+        final Editable text = getText();
+        final Layout layout = getLayout();
+        for (int i = 1; i < count; i++) {
+            if (text.charAt(layout.getLineStart(i) - 1) == '\n') {
+                number++;
+                y = layout.getLineTop(i);
```

Comment:
How does this look in todo.txt where we add extra padding to the top of the line?

Reply:
Here is the screenshot, not bad.

![Screenshot_20230708_161105_net.gsantner.markor_test.jpg](https://github.com/gsantner/markor/assets/60415467/e8a1b8b5-d54c-42d4-93d5-6bea9e7aa81f)

Reply:
with the todo.txt improvement I pushed 
![screenshot-2023-07-09__20-29-36](https://github.com/gsantner/markor/assets/6735650/95c18774-3afd-4390-8dc7-475e5ac80cae)

Reply:
It looks better 😊